### PR TITLE
PIM-3664: Product media stacktrace regression

### PR DESCRIPTION
**Pull request by solivier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Yes
| New feature?         | No
| BC breaks?           | No
| CI currently passes? | Yes
| Tests pass?          | Yes
| Scenarios pass?      | Yes
| Checkstyle issues?*  | No
| PMD issues?**        | No
| Changelog updated?   | Yes
| Fixed tickets        | PIM-3664
| DB schema updated?   | No
| Migration script?    | No
| Doc PR               | -

### CODE REVIEW

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 1]
```diff
@@ -60,11 +62,15 @@ public function process($item)
...
-                ['field_name' => 'media', 'prepare_copy' => true]
-            );
+            try {
+                $data['media'] = $this->serializer->normalize(
+                    $item->getMedia(),
+                    'flat',
+                    ['field_name' => 'media', 'prepare_copy' => true]
+                );
+            } catch (FileNotFoundException $e) {
+                throw new InvalidItemException($e->getMessage(), ['item' => $item->getMedia()[0]->getValue()]);
```
Comment (solivier) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Definitely not sure about that line for item element

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php [hunk 2]
```diff
@@ -223,10 +228,6 @@ public function createFromFilePath($filePath)
      */
     public function getExportPath(ProductMediaInterface $media)
     {
-        if (null === $this->getFilePath($media)) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: Should we not throw an exception here ?

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php [hunk 3]
```diff
@@ -334,22 +335,26 @@ protected function write($filename, $content, $overwrite = false)
      *
      * @param ProductMediaInterface $media
      *
-     * @return string|null the path of the media or null if the media has no file attached
+     * @return string the path of the media or null if the media has no file attached
```
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: yo actually return null 10 lines under this line

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php [hunk 4]
```diff
@@ -334,22 +335,26 @@ protected function write($filename, $content, $overwrite = false)
...
      * @throws FileNotFoundException in case the file of the media does not exist or is not readable
      */
     public function getFilePath(ProductMediaInterface $media)
     {
-        if ($this->fileExists($media)) {
-            $path = $this->uploadDirectory . DIRECTORY_SEPARATOR . $media->getFilename();
-            if (!is_readable($path)) {
-                throw new FileNotFoundException($path);
-            }
+        if (null === $media->getFilename()) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: This situation is not possible

Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: in fact it's used during mass edit, and it does not come from DB. When you want to unset all images on several products

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php (hunk 3 above)
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: right

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php (hunk 2 above)
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: this method only generate the export path, and doesn't care if file exists or not, exceptions will be raised during copy or other method that really manipulate media.

File: spec/Pim/Bundle/CatalogBundle/Manager/MediaManagerSpec.php [hunk 5]
```diff
@@ -176,4 +176,26 @@ function it_can_create_media_from_the_file_path($factory, ProductMediaInterface
...
+    ) {
+        $this->shouldThrow(new \LogicException('File path should not be null'))->during(
+            'duplicate',
+            [$source, $target, 'publish']
+        );
+    }
+
+    function it_returns_false_if_media_does_not_exists_during_copy(ProductMediaInterface $media)
+    {
+        $media->getFilename()->shouldBeCalled()->willReturn('/tmp/invalid');
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/CatalogBundle/Manager/MediaManagerSpec.php]: We should really not do that :/

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php (hunk 4 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: but you throw an exception 5 lines below. I don't understand

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 6]
```diff
@@ -31,26 +31,28 @@ class ProductToFlatArrayProcessor extends AbstractConfigurableStepElement implem
...
     /**
      * @param Serializer     $serializer
      * @param ChannelManager $channelManager
+     * @param string         $uploadDirectory
      */
     public function __construct(
         Serializer $serializer,
-        ChannelManager $channelManager
+        ChannelManager $channelManager,
+        $uploadDirectory
```
Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: please add this BC break in the changelog

**Pull request merged**